### PR TITLE
Working Cisco IOS macro editing (fixes #55211)

### LINKED ISSUE #55211: IOS: configuring macro fails
##### SUMMARY
Configuring a Cisco IOS macro through ios_config fails. I've tried various variations as found while searching the web and checked out #38420. None got me a working playbook.
(I've come up with a fix to edit_macro() for which I'll open a PR)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ios_config
plugins/cliconf/ios.py  : edit_macro()

##### ANSIBLE VERSION

```paste below
ansible 2.7.8
  config file = /home/ops/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible/plugins/modules', '/home/ops/local/ansible-modules']
  ansible python module location = /home/ops/lib/python3.7/site-packages/ansible
  executable location = /home/ops/bin/ansible
  python version = 3.7.0 (default, Jan 16 2019, 12:24:08) [GCC 4.4.7 20120313 (Red Hat 4.4.7-17)]

```

##### CONFIGURATION

```paste below
DEFAULT_ACTION_PLUGIN_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible_plugins/action_plugins', '/home/ops/local/ansible-plugins/action']
DEFAULT_ASK_VAULT_PASS(/home/ops/ansible/ansible.cfg) = True
DEFAULT_FILTER_PLUGIN_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible_plugins/filter_plugins', '/home/ops/local/ansible-plugins/filter']
DEFAULT_FORKS(/home/ops/ansible/ansible.cfg) = 20
DEFAULT_GATHERING(/home/ops/ansible/ansible.cfg) = explicit
DEFAULT_HOST_LIST(/home/ops/ansible/ansible.cfg) = ['/home/ops/ansible/inventoryyaml']
DEFAULT_INVENTORY_PLUGIN_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible_plugins/inventory_plugins', '/home/ops/local/ansible-plugins/inventory']
DEFAULT_LOAD_CALLBACK_PLUGINS(/home/ops/ansible/ansible.cfg) = True
DEFAULT_LOOKUP_PLUGIN_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible_plugins/lookup_plugins', '/home/ops/local/ansible-plugins/lookup']
DEFAULT_MODULE_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible/plugins/modules', '/home/ops/local/ansible-modules']
DEFAULT_STDOUT_CALLBACK(/home/ops/ansible/ansible.cfg) = yaml
DEFAULT_TRANSPORT(/home/ops/ansible/ansible.cfg) = local
DEFAULT_VARS_PLUGIN_PATH(/home/ops/ansible/ansible.cfg) = ['/usr/share/ansible_plugins/vars_plugins', '/home/ops/local/ansible-plugins/vars']
INVENTORY_ENABLED(/home/ops/ansible/ansible.cfg) = ['yaml']
PERSISTENT_COMMAND_TIMEOUT(/home/ops/ansible/ansible.cfg) = 120
PERSISTENT_CONNECT_RETRY_TIMEOUT(/home/ops/ansible/ansible.cfg) = 60
PERSISTENT_CONNECT_TIMEOUT(/home/ops/ansible/ansible.cfg) = 120
RETRY_FILES_ENABLED(/home/ops/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

RedHat 6.8
IOS-XE version 03.06.08.E (15.2-2e8)

##### STEPS TO REPRODUCE

Try to configure a macro with at least 15-20 lines. I'm not sure if the number of lines is an issue, but my 802.1x port macro consistently failed and had at least 15 lines.

```yaml
    - name: "Configure {{ macro_name }} macro"
      ios_config:
        lines: "{{ dot1x_enforce_macro }}"
        parents: "macro name {{ macro_name }}"
        after: '@'
        match: line
        replace: block
```

##### EXPECTED RESULTS

Macro configured

##### ACTUAL RESULTS

ios_config fails with error message that the macro exceeds the number of allowed characters.

( tried with latest devel branch edit_macro() )

```paste below
fatal: [myswitch.domain.tld]: FAILED! => changed=false
  module_stderr: |-
    /home/user/.ansible/tmp/ansible-tmp-1555087469.0910754-193292457425993/AnsiballZ_ios_config.py:17: DeprecationWarning: the imp module is deprecated in favour of impo
      import imp
    Traceback (most recent call last):
      File "/home/user/.ansible/tmp/ansible-tmp-1555087469.0910754-193292457425993/AnsiballZ_ios_config.py", line 113, in <module>
        _ansiballz_main()
      File "/home/user/.ansible/tmp/ansible-tmp-1555087469.0910754-193292457425993/AnsiballZ_ios_config.py", line 105, in _ansiballz_main
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
      File "/home/user/.ansible/tmp/ansible-tmp-1555087469.0910754-193292457425993/AnsiballZ_ios_config.py", line 48, in invoke_module
        imp.load_module('__main__', mod, module, MOD_DESC)
      File "/usr/local/lib/python3.7/imp.py", line 235, in load_module
        return load_source(name, filename, file)
      File "/usr/local/lib/python3.7/imp.py", line 170, in load_source
        module = _exec(spec, sys.modules[name])
      File "<frozen importlib._bootstrap>", line 630, in _exec
      File "<frozen importlib._bootstrap_external>", line 728, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/tmp/ansible_ios_config_payload_rp2c55k8/__main__.py", line 543, in <module>
      File "/tmp/ansible_ios_config_payload_rp2c55k8/__main__.py", line 474, in main
      File "/tmp/ansible_ios_config_payload_rp2c55k8/__main__.py", line 333, in edit_config_or_macro
      File "/tmp/ansible_ios_config_payload_rp2c55k8/ansible_ios_config_payload.zip/ansible/module_utils/connection.py", line 176, in __rpc__
    ansible.module_utils.connection.ConnectionError: 'Cliconf' object has no attribute 'check_edit_config_capabiltiy'
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1

```

### PR DESCRIPTION
**Pull request by ajsiersema**
##### SUMMARY

Working implementation for editing Cisco IOS macros
Fixes #55211

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ios_config

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by pabelanger** [comment]
recheck

**Comment by pabelanger** [comment]
We should be able to write a test case for this, in tests/integration/targets/ios_config.

**Comment by pabelanger** [comment]
The ios-python27 failure is know issue, we are working to fix it.

**Comment by justjais** [comment]
recheck

**Comment by justjais** [comment]
@ajsiersema can u plz write a test WRT to this fix in ios integration tests before we can go ahead and merge the PR. Also, I think you'll need to rebase the branch as well for changes to go in `2.9` i.e. devel.

**Comment by ajsiersema** [comment]
@justjais : I see the build passed testing. Are these only python tests , i.e. extra ios integration tests are still needed ?
I'll look into a 2.9 rebase

**Comment by justjais** [comment]
@ajsiersema You'll have to write new integration tests for the respective use case in ios_config integration test and it's not python based it's similar to the ansible play you actually run with assert to test and verify the output of the play (ref: [ios_config integration tests](https://github.com/ansible/ansible/tree/devel/test/integration/targets/ios_config/tests/cli)).

Also, rebase is not needed as it's already in `devel`

**Review by justjais — COMMENTED** [review]
@ajsiersema Needs integration tests for the added use case.

**Comment by ajsiersema** [comment]
> @ajsiersema Needs integration tests for the added use case.

I've create integration tests, now working on getting these into the PR

**Comment by ajsiersema** [comment]
Added integration tests and tested those against the latest devel. Still failed on our switches.

E.g.

**macro.j2**
```
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
```

Running ```ansible-test network-integration --color no --local --inventory ~/inventory.networking --testcase macro --tags="cli" ios_config```

results in:

```
TASK [ios_config : Define macro name]
  ok: [testswitch]
TASK [ios_config : setup - remove configuration]
  changed: [testswitch]
TASK [ios_config : configure macro]
  changed: [testswitch]
TASK [ios_config : Check if macro has been added]
ok: [testswitch] => changed=false
  msg: All assertions passed
TASK [ios_config : configure macro again - idempotent check]
  changed: [testswitch]
TASK [ios_config : macro already/correctly configured ?]
  fatal: [testswitch]: FAILED! => changed=false
    assertion: result.changed == false
    evaluated_to: false
    msg: Assertion failed
```

and a running switch config:

```
macro name MACRO_ANSIBLE_TEST

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
@

```


Adding some more bytes to **macro.j2** :
```
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
 H12345678 123456789 123456789 123456789 123456789 123456789 123456789
```

results in:

```
TASK [ios_config : Define macro name]
  ok: [testswitch]
TASK [ios_config : setup - remove configuration]
  changed: [testswitch]
TASK [ios_config : configure macro]
  An exception occurred during task execution. To see the full traceback, use -vvv. The error was: mySwitch(config)#
  fatal: [testswitch]: FAILED! => changed=false
    module_stderr: |-
      Traceback (most recent call last):
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 139, in <module>
          _ansiballz_main()
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 131, in _ansiballz_main
          invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 68, in invoke_module
          imp.load_module('__main__', mod, module, MOD_DESC)
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 569, in <module>
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 500, in main
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 362, in edit_config_or_macro
        File "/tmp/ansible_ios_config_payload_8umhKB/ansible_ios_config_payload.zip/ansible/module_utils/connection.py", line 186, in __rpc__
      ansible.module_utils.connection.ConnectionError: 
      % Macro length exceed its limit of 3000 bytes. Truncating the macro.
      mySwitch(config)#
      mySwitch(config)#config terminal
      config terminal
             ^
      % Invalid input detected at '^' marker.
      mySwitch(config)#
    module_stdout: ''
    msg: |-
      MODULE FAILURE
      See stdout/stderr for the exact error
    rc: 1
```

and running switch config:

```
macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
 H12345678 123456789 123456789 123456789 123456789 123456789 123456789
@

```

**Comment by ajsiersema** [comment]
Obviously with the changes in this PR, the integration tests will pass :)

**Comment by justjais** [comment]
recheck

**Comment by justjais** [comment]
@ajsiersema Thanks for pushing the integration tests, but currently the tests are failing coz of: https://logs.zuul.ansible.com/20/55220/de32d995cd7394e72b62127e19cc6d73cc2ffed7/third-party-check/ansible-test-network-integration-ios-python37/b5284bd/controller/ara-report/file/02c40892-71db-4e33-81f6-6f56703ccde9/#line-9 (ref: https://logs.zuul.ansible.com/20/55220/de32d995cd7394e72b62127e19cc6d73cc2ffed7/third-party-check/ansible-test-network-integration-ios-python37/b5284bd/controller/ara-report/).
Kindly, resolve the error before we can merge the PR.

**Comment by hexdump0x0200** [comment]
@ajsiersema, @justjais Error in integration tests looks as if the device on which they run does not support macros. This is quite possible even for new devices. I have some quite modern devices that do not support macros. There are even two devices of different hardware platforms running on the same IOS version, with one device supporting macros, and the second is not.

The proposed PR works on my equipment and integration tests are also completed successfully.

**Comment by justjais** [comment]
@hexdump0x0200 can you plz let me know the version of ios on which you are able to run the respective integration test and version on which it fails.

**Comment by ajsiersema** [comment]
> @hexdump0x0200 can you plz let me know the version of ios on which you are able to run the 
> respective integration test and version on which it fails.

I successfully ran the integration tests on these switches and software versions :
9407R IOS XE Version 16.9.03
4503E IOS XE Version 03.06.08.E.152-2.E8
3560CG IOS Version 15.0(2)SE12
3750X IOS Version 15.0(2)SE11

**Comment by ajsiersema** [comment]
Do we need a conditional check in the integration test playbook ?

If the PR + checks succeed, do I need to create changelog fragments and backports to e.g. 2,8 for inclusion or will the ansible networking team take care of getting the code into stable releases ?

**Comment by justjais** [comment]
@ajsiersema Yes, since lab IOS version is `15.6(2)T` it's better to have a conditional statement to run respective tests for the macro supported IOS version.

Also, once this PR is merged in `devel` then u can create a backport for the fix under `stable-2.8` with changelog for the backport PR.

**Comment by ajsiersema** [comment]
I've added a check for macro support. This will however skip all edit_macro() integration checks in the current lab environment. Not quite sure if that should be the end result. I'd rather have an IOS lab environment on which these integration tests are automatically running as well.

**Comment by hexdump0x0200** [comment]
For now it looks like a good solution. Macros are most in demand on the switches, it seems that the vendor is not in a hurry to implement this function on routers (integration tests are performed on a virtual router). However, I found an old router with IOS 15.1 which supports macros, so maybe it's a matter of time when support appears in branch 15.6.

**Comment by hexdump0x0200** [comment]
The "when" condition for the block at the very end is very poorly read. It’s not even immediately clear to what it refers: to the block or to the last task from the block, one has to look closely at the indents.
So it would be much better:
```yaml
- name: "ios_config macro integration tests"
  when: supports_macro is succeeded
  block:
...
```

**Comment by ajsiersema** [comment]
> 
> 
> The "when" condition for the block at the very end is very poorly read. It’s not even immediately clear to what it refers: to the block or to the last task from the block, one has to look closely at the indents.
> So it would be much better:
> 
> ```yaml
> - name: "ios_config macro integration tests"
>   when: supports_macro is succeeded
>   block:
> ...tasks
> # end block "ios_config macro integration tests"
> - name: after-block task
> ...
> ```

I check this file for the coding style:
test/integration/targets/ios_smoke/tests/cli/common_config.yaml

However, I fully agree with you with regard to the readability, my own playbooks also have the when at the top :) Will resubmit with **when** moved near **block**

**Comment by jamiereid** [comment]
Hi Everyone,

I've hit the same error message in #55211 (@ajsiersema interestingly enough, the macro I am trying to create is also for DOT1X)

Is there some work I could maybe do to help get this PR over the line?

**Comment by ajsiersema** [comment]
> I've hit the same error message in #55211 (@ajsiersema interestingly enough, the macro I am trying to create is also for DOT1X)
> Is there some work I could maybe do to help get this PR over the line?

Thanks for the feedback.
As soon as the PR is accepted, I don't mind getting submitting the PR for 2.8 / 2.9. First, someone from the Ansible network team needs to accept the PR.
@hexdump0x0200  , @justjais  : is there anything else we can/need to do ?

File: test/integration/targets/ios_config/tests/cli/macro.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      replace: block
+      lines: "{{ lookup('template', 'basic/macro.j2') }}"
+    register: result
+
+  - name: "macro already/correctly configured ?"
+    assert:
+      that:
+        - "result.changed == false"
+                    
+  - name: "teardown"
```
Comment (justjais) [inline test/integration/targets/ios_config/tests/cli/macro.yaml]: @ajsiersema Can u move the teardown play to `always` block, which will ensure that even if the play fails in between teardown will run always.
```suggestion
  always:
  - name: "teardown"
```

**Review by justjais — COMMENTED** [review]
@ajsiersema apologies for the delay in review, somehow this slipped my track. 
PR changes lgtm, but I have commented for a minor cosmetic update in the integration test.

File: test/integration/targets/ios_config/tests/cli/macro.yaml (hunk 1 above)
Comment (ajsiersema) [reply test/integration/targets/ios_config/tests/cli/macro.yaml]: @justjais : thanks for the review, just pushed the change (I was bogged down myself for a few days, common cold, busy schedule, etc)

**Review by justjais — APPROVED** [review]

**Pull request merged**